File: js/047_PermutationsII.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permuteUnique = function(nums) {
    nums.sort( function(a,b){return a-b;} );
    var ret = [];
    backtrace(nums.concat(),0,ret);
    return ret;
};

var swap = function(nums,i,j){
    var t = nums[i];
    nums[i] = nums[j];
    nums[j] = t;
};

var backtrace = function(nums, index, ret){
    if(index == nums.length - 1 ){
        ret.push(nums.concat());
        return;
    }
    for(var i = index; i < nums.length; ++i){
        if(i > index && nums[i] === nums[index]) continue;
        swap(nums,index,i);
        backtrace(nums.concat(),index+1,ret);
    }
};

File: cpp/047_PermutationsII.cpp
class Solution {
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        std::sort(nums.begin(), nums.end());
        vector<vector<int>> ret;
        backtrack(nums,0,ret);
        return ret;
    }
    
private:
    void backtrack(vector<int> nums, int index, vector<vector<int>>& ret){
        if(index == nums.size()-1){
            ret.push_back(nums);
            return;
        }
        for(int i = index; i < nums.size(); ++i){
            if(index != i && nums[index] == nums[i]) continue;
            swap(nums[index],nums[i]);
            backtrack(nums,index+1,ret);
        }
    }
};